### DISCUSSION TITLE
Replace <a> with <a> breaks the link

### DISCUSSION THREAD

**Discussion by hogoww** [remarkjs/remark#1369] [Q&A] [answered]
Hello,

I'm trying to change links that come from a random source to local links.
For context, I am trying to avoid reloads when I click on a local link, and redirect it to /app/whatever/url 

Basic V1: 
```
<Markdown> {markdown} </Markdown>
```
This works, but as the link comes from users inputs (basically), it triggers a full reload even if everything is loaded.

V2:
```
<Markdown
          components={{
              a: props => { return <a href={props.href}>{props.children}</a> }
          }}>{markdown}</Markdown>
```
This renders something that looks like a link, but does not navigate.
On top of V2, I tried the following which all gave similar results:
- Adding an onClick on the <a> which never triggers;
- Adding remark-gfm;
- Using the syntax a(props){...};
- Using react-router-dom Links;
- Hacked the framework to check that my url was not sanitize
Searching the internet gave me multiple solution, but I feel like I'm missing something about how things work under the hood.
And I'm running out of ideas :D

Does anyone have an idea of what I'm missing please ? =)

**Comment by ChristianMurphy** [comment]
Welcome @hogoww! 👋 
Some questions to help better help you.

> I am trying to avoid reloads

One of the main advantages of React is that it manages when you reload, and generally you shouldn't need to worry about a reload it will be fast (exceptions of course for infinite reloads, compute heavy workloads, and blocking processes).
I don't hear any of those exceptions in your message, so my initial answer would be stop worrying and let React do it's thing 🙂 

> I click on a local link, and redirect it to /app/whatever/url

What do you consider a "local link". Do you mean an absolute path on the same domain? A path relative to the current path? An anchor within the current page? Or something else entirely based off of one of React/Next's many routing systems?

> This works, but as the link comes from users inputs (basically), it triggers a full reload even if everything is loaded.

I'm not sure I understand your comment.
If you go to another page, it needs to reload, that's how links work.
Are you using something more custom that overrides how URLs are handled? Like a SPA style router and offline app caching? If so have you considered that it could be a bug in how you have that router configured?

> Searching the internet gave me multiple solution, but I feel like I'm missing something about how things work under the hood.
> And I'm running out of ideas :D

Reoccuring theme here, solutions to what?
You are a bit unclear on what your actual problem is or what you've actually tried with the libraries you list.
Some helpful reading:
* https://xyproblem.info/
* https://stackoverflow.com/help/minimal-reproducible-example

---

The community is happy to help.
Help us help you, by taking some time to articulate what you are trying to solve a bit more clearly. 🙂

**Reply by hogoww** [reply]
> Welcome @hogoww! 👋 Some questions to help better help you.
👋 
> > I am trying to avoid reloads
> 
> One of the main advantages of React is that it manages when you reload, and generally you shouldn't need to worry about a reload it will be fast (exceptions of course for infinite reloads, compute heavy workloads, and blocking processes). I don't hear any of those exceptions in your message, so my initial answer would be stop worrying and let React do it's thing 🙂
Indeed, however, I am trying to provide navigation onto an interactive application, which means that everything is already loaded, and it shouldn't trigger a refresh but change the focus.

I'll explain with screenshot, hopefully, that'd be more understandable.
Here, the focus is on a map tip, I have a markdown text with a link "patrol 25".
The patrol 25 in question is the green container right next to the exclamation point.
![image](https://github.com/user-attachments/assets/fd94929f-555a-4ee2-82ef-c0d95967392b)
I want that when I click on the link in the markdown part, it changes the focus onto the pack next to it.
![image](https://github.com/user-attachments/assets/d282f6e2-8f0e-4986-a89b-8133f48e26bf)
As the markdown is provided by users, it won't have a local link, but a regular one, see explanation bellow.


> > I click on a local link, and redirect it to /app/whatever/url
> 
> What do you consider a "local link". Do you mean an absolute path on the same domain? A path relative to the current path? An anchor within the current page? Or something else entirely based off of one of React/Next's many routing systems?

I think I'm explaining myself poorly with the wrong vocabulary, I'll try with a more concrete example.
My website URL is https://deplete.fyi.
When a user submits a markdown, it'll be with a link such as https://deplete.fyi/app/someNumber.
The markdown sees a full link, and reload the application.
I wanted to use what I call a "local" url, which react-router-dom would catch and reload "/app/someNumber".
Basically, just removing the domain (https://deplete.fyi/)
Here's an exerpt of my router, in case that helps:
```
<BrowserRouter>
   <Routes>
      <Route path="/app/:entityId" element=<ResolvePathOntoMap />
      </Route>
   </Routes>
</BrowserRouter>
```

> > This works, but as the link comes from users inputs (basically), it triggers a full reload even if everything is loaded.
> 
> I'm not sure I understand your comment. If you go to another page, it needs to reload, that's how links work. Are you using something more custom that overrides how URLs are handled? Like a SPA style router and offline app caching? If so have you considered that it could be a bug in how you have that router configured?

I think I have already answered in a previous answer, but I'll try again, maybe it'll give more details =).
I'm trying to provide an interactive application with the interaction that changes the URL, to support navigation more seemlessly.
Therefore you click on part of the application, it doesn't reload, but changes (Navigate in react-router-dom)the current URL.
You click another part, it updates the URL again.
You click on previous, you're back to the previous part of the application.
> 
> > Searching the internet gave me multiple solution, but I feel like I'm missing something about how things work under the hood.
> > And I'm running out of ideas :D
> 
> Reoccuring theme here, solutions to what? You are a bit unclear on what your actual problem is or what you've actually tried with the libraries you list. Some helpful reading:
> 
> * https://xyproblem.info/
> * https://stackoverflow.com/help/minimal-reproducible-example
> 
> The community is happy to help. Help us help you, by taking some time to articulate what you are trying to solve a bit more clearly. 🙂

I'm mostly implementing programming languages, so I think that I express myself poorly when it comes to web dev.
I thought it was better explained that what you read, sorry ^^"
Does that make more sense ? =)
Regardless, thanks for the initial answer!

**Reply by ChristianMurphy** [reply]
Thanks for the additional context @hogoww!

Taking a step back, this isn't so much an MDX question.
You've figured out how to add elements to the page.
This is more a "How do I get React to focus on things inside a Canvas?"
Which isn't really the purview of the MDX discussion forum, but I'll offer some pointers anyway. 🙂 

---

> When a user submits a markdown, it'll be with a link such as https://deplete.fyi/app/someNumber.
> The markdown sees a full link, and reload the application.
> I wanted to use what I call a "local" url, which react-router-dom would catch and reload "/app/someNumber".
> Basically, just removing the domain (https://deplete.fyi/)
> Here's an excerpt of my router, in case that helps:
>
> ```jsx
> <BrowserRouter>
>    <Routes>
>       <Route path="/app/:entityId" element=<ResolvePathOntoMap />
>       </Route>
>    </Routes>
> </BrowserRouter>
> ```

I still don't really understand what you are trying to get at here.
Whether plain browser routing, or using React-Router and the navigation API, you can't (directly) link to something inside a canvas.
You can only link to pages or DOM elements on the page with a uniquely identified anchor.

Are you saying that the internal link includes a unique identifying number, then that number is passed to the component, and based off that number the component knows how to animate the canvas?

> I'm trying to provide an interactive application with the interaction that changes the URL, to support navigation more seemlessly.
> Therefore you click on part of the application, it doesn't reload, but changes (Navigate in react-router-dom)the current URL.
> You click another part, it updates the URL again.
> You click on previous, you're back to the previous part of the application.

If you want React-Router to update the URL without the page navigating, you need to use their `<Link>` element https://reactrouter.com/en/main/components/link

**Reply by hogoww** [reply]
> Thanks for the additional context @hogoww!

Thank you !

> Taking a step back, this isn't so much an MDX question. You've figured out how to add elements to the page. This is more a "How do I get React to focus on things inside a Canvas?" Which isn't really the purview of the MDX discussion forum, but I'll offer some pointers anyway. 🙂

Actually, I already have this behavior working in multiple other components, and I kinda had to redo parts of react-router-dom.
(Took me a while to understand properly though...)
It's really just when I try to apply the same principle on react-markdown, that I can't figure out.
I just need to get the click event on the link to be honest, but simply using the `components` arguments breaks the click event, as far as I can tell :/

That's actually why I was trying to understand the initial code snippet difference, and why the link works in V1 & not in V2.
That's just the minimal reproducible example that I try to work with to figure it out.

> Are you saying that the internal link includes a unique identifying number, then that number is passed to the component, and based off that number the component knows how to animate the canvas?

Basically, yes I am!

**Reply by ChristianMurphy** [reply]
> That's actually why I was trying to understand the initial code snippet difference, and why the link works in V1 & not in V2.
> That's just the minimal reproducible example that I try to work with to figure it out.

@hogoww what you shared is a snippet.
When I try to reproduce it, it works as expected https://stackblitz.com/edit/github-kt2wuv?file=src%2Fapp.tsx

I need more context on what you saw, that didn't work, to offer a meaningful suggestion.

> breaks the click event, as far as I can tell

@hogoww I can't peer into your mind 😅
What click event? You haven't shared any.

**Reply by hogoww** [reply]
> > That's actually why I was trying to understand the initial code snippet difference, and why the link works in V1 & not in V2.
> > That's just the minimal reproducible example that I try to work with to figure it out.
> 
> @hogoww what you shared is a snippet. When I try to reproduce it, it works as expected https://stackblitz.com/edit/github-kt2wuv?file=src%2Fapp.tsx
> 
> I need more context on what you saw, that didn't work to offer a meaningful suggestion.
> 

That is true...
I'll check on my working environment again.
I was thinking it could be related to CSS, but I'm very bad at CSS, so I'll compare with your poc :D

> > breaks the click event, as far as I can tell
> 
> @hogoww I can't peer into your mind 😅 What click event? You haven't shared any.

I'm talking about the implicit click event that is given to a `<a>`  tag. 
I tried to add it explicitly too "Adding an onClick on the which never triggers;"
Is that clearer ^^"

**Reply by ChristianMurphy** [reply]
> I'm talking about the implicit click event that is given to a balise.

What is "balise"?
Is that a framework? A library? A typo?

**Reply by hogoww** [reply]
> > I'm talking about the implicit click event that is given to a balise.
> 
> What is "balise"? Is that a framework? A library? A typo?

*cough* It's an untranslated word, and an unexpected formating :D
Here's a better worded phrase, hopefully: 
I'm talking about the implicit click event that is given to `<a>` *tag*.

**Reply by ChristianMurphy** [reply]
> implicit click event that is given to `<a>` *tag*.

What do you mean by an implicit click?
Plain browser DOM and React don't add listeners unless explicitly asked to.
Do you have jQuery or another DOM listening library added?

What is it that you want this click to do, or not do?

Also I thought you were leveraging react-router? Which manipulates the URL rather than doing a stateful click handler.

**Reply by hogoww** [reply]
I did something similar to your poc, by just doing both version side to side.
First one (aka V1) reacts to a click on the link, and updates the page.
Second one (aka V2) doesn't react to the click on the link, and does not update the page.
The generated HTML look exactly alike to me (which makes sense).
The CSS look exactly alike aswell.
I have no clue why I have this problem locally.
Do you think it could be a CSS issue? 
If so, I'll try to update the CSS till it works.

![image](https://github.com/user-attachments/assets/2039230f-1baf-4621-9eaf-f8f64785716a)

**Reply by hogoww** [reply]
> > implicit click event that is given to `<a>` _tag_.
> 
> What do you mean by an implicit click? Plain browser DOM and React don't add listeners unless explicitly asked to. Do you have jQuery or another DOM listening library added?
> 
> What is it that you want this click to do, or not do?
> 
> Also I thought you were leveraging react-router? Which manipulates the URL rather than doing a stateful click handler.

It's not important, and we should probably drop that part of the conversation, since I cannot explain properly :/
I'm trying to talk about the basic HTML runtime behavior for a `<a>` tag
I don't think that's very relevant ^^"

**Reply by ChristianMurphy** [reply]
> Do you think it could be a CSS issue?
> If so, I'll try to update the CSS till it works.

This does not sound like a CSS issue.
CSS changes appearance, but not behavior.

> First one (aka V1) reacts to a click on the link, and updates the page.
> Second one (aka V2) doesn't react to the click on the link, and does not update the page.

My gut sense is something is messed up either with:
1. A third-party library on the page, either analytics (like Google Analytics or Matomo) or a really old UI library that takes over all elements was added (like some kind of jQuery plugin).
2. Something is configured in your build tool that treats links different
3. You have a bug in your router that breaks URLs randomly and it has happens to show up mostly when you click the second link

For something more specific, _please_ **please** **_please_**, create a minimal reproducible example that shows the issue you are seeing. https://stackoverflow.com/help/minimal-reproducible-example
In a sandbox like https://codesandbox.io/p/sandbox/github/remarkjs/.github/tree/main/sandbox-templates/react-markdown-with-vite or https://codesandbox.io/p/sandbox/github/remarkjs/.github/tree/main/sandbox-templates/react-markdown-with-vite
That people can see, and click through, and debug.

**Reply by hogoww** [reply]
> > Do you think it could be a CSS issue?
> > If so, I'll try to update the CSS till it works.
> 
> This does not sound like a CSS issue. CSS changes appearance, but not behavior.
> 

Thanks for the sanity check !

> > First one (aka V1) reacts to a click on the link, and updates the page.
> > Second one (aka V2) doesn't react to the click on the link, and does not update the page.
> 
> My gut sense is something is messed up either with:
> 
> 1. A third-party library on the page, either analytics (like Google Analytics or Matomo) or a really old UI library that takes over all elements was added (like some kind of jQuery plugin).
> 2. Something is configured in your build tool that treats links different
> 3. You have a bug in your router that breaks URLs randomly and it has happens to show up mostly when you click the second link
> 
> For something more specific, please please please, create a minimal reproducible example that shows the issue you are seeing. https://stackoverflow.com/help/minimal-reproducible-example In a sandbox like https://codesandbox.io/p/sandbox/github/remarkjs/.github/tree/main/sandbox-templates/react-markdown-with-vite or https://codesandbox.io/p/sandbox/github/remarkjs/.github/tree/main/sandbox-templates/react-markdown-with-vite That people can see, and click through, and debug.

I'm trying this atm !
The only differrence so far, is that you backquote your links.

**Comment by hogoww** [comment]
Small update, I was able to make it work locally, outside of a canvas.
Seems to be canvas related.
So I guess this conversation could be close !
I'll update still if/when I find the weird interaction.
Thanks a lot for your help @ChristianMurphy !

**Comment by hogoww** [comment]
Hello,
After way too much debug, I was able to recreate a small reproducible case: https://stackblitz.com/edit/github-kt2wuv-mswjdj?file=src%2Fapp.jsx
It might be reducible a little more though! 
I find it quite convulated.
Is that expected behavior? 
To me, it looks like a bug, I'm just not sure of which framework tbh :D
I feel like it should be reported somewhere, but where...

**Reply by ChristianMurphy** [reply]
I don't really understand this MMR.
What do you _want_ it to do?
What is it doing instead?

---

Making an educated guess, you want to instrument the link to also run some custom function?
If so use `onClick` on the component itself.

```jsx
import Markdown from 'react-markdown';

const markdown = `
**Pluto** (minor-planet designation: *134340 Pluto*)
is a
[dwarf planet](https://en.wikipedia.org/wiki/Dwarf_planet)
in the
[Kuiper belt](https://en.wikipedia.org/wiki/Kuiper_belt).
`;

const eventHandler = (event) => {
  // if you don't want the link to be followed preventDefault
  // event.preventDefault();
  console.log('event called');
};

function App() {
  return (
    <Markdown
      components={{
        a: (props) => {
          return (
            <a onClick={eventHandler} href={props.href}>
              {props.children}
            </a>
          );
        },
      }}
    >
      {markdown}
    </Markdown>
  );
}

export default App;
```

https://stackblitz.com/edit/github-kt2wuv-8nt11n?file=src%2Fapp.jsx

**Reply by hogoww** [reply]
This started as a question as of why I had a bug.This bug is now reproducible with the link from the [previous post](https://stackblitz.com/edit/github-kt2wuv-mswjdj?file=src%2Fapp.jsx)

Very basically, I want the following:
When I click on a `<a href="...">`, I want it to open a new page (which is literally the default behavior of the `<a>` AFAIK)
As you can see in the POC, it does work without override. 
It does not work when you override.

Now, I know how to fix.
However, I think this is a bug in a framework.
I do not know whether the bugged framework is react, react-markdown, or even something else.
I'd like to report that bug, as it took a lot of work to find the exact root cause.

**Reply by ChristianMurphy** [reply]
> I want it to open a new page (which is literally the default behavior of the <a> AFAIK)

It is not, same page is the default https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#target

**Reply by hogoww** [reply]
Fair enough.
Still, in the POC, the anchor does not receive any event, so you cannot click on it.

**Reply by ChristianMurphy** [reply, marked as answer]
>  Still, in the link, the anchor does not receive any event, so you cannot click on it.

There's a lot going wrong here, and there is no framework bug, but there is a good learning opportunity.

---

The main issue you are running into is order of operations (mousedown before click) and stable identifiers.

Your event handler updates state, which will re-render the component
```jsx
  const [contextMenu, setContextMenu] = useState({});

  const func = () => {
    setContextMenu({});
  };
```

In the second example you define the `a` function has an anonymous inline function

```jsx
   <Markdown
        components={{
          a: (props) => { // <= this is anonymous inline
            return <a href={props.href}>{props.children}</a>;
          },
        }}
      >
```
which means the `a`/link component will be re-defined every time the page reloads.

Because `mousedown` happens before click https://developer.mozilla.org/en-US/docs/Web/API/Element/mousedown_event

```jsx
document.addEventListener('mousedown', func);
```

The link you are trying to click on disappears before you can actually click it.
That's why you see a little render flash on the second example.

to fix this, define `components` outside your component.
```jsx
const components = {
  a: (props) => {
    return <a href={props.href}>{props.children}</a>;
  },
};
```

then reference it in your render function

```jsx
<Markdown components={components}>{markdown}</Markdown>
```

this gives a stable reference for the `a` component which react doesn't need to replace.

working example: https://stackblitz.com/edit/github-kt2wuv-zfjlkb?file=src%2Fapp.jsx

---

In addition you have several other bugs, including not cleaning up your event listeners on re-render and not having the function inside the effect.
If you choose to keep a document wide listener it should look like:

```jsx
  useEffect(() => {
    const eventHandler = () => {
      setContextMenu({});
    };
    document.addEventListener('mousedown', eventHandler);
    return () => {
      document.removeEventListener('mousedown', eventHandler) // cleanup listener
    };
  }, [setContextMenu]); // watch for changes to state function
```

see:
* https://blog.logrocket.com/understanding-react-useeffect-cleanup-function/
* https://react.dev/reference/react/useEffect#removing-unnecessary-function-dependencies

---

That said you probably don't want to add a listener to the whole page.
That goes against react convention, scope it to your component instead and you can avoid needing the effect in the first place

```jsx
function App() {
  const [contextMenu, setContextMenu] = useState({});

  const eventHandler = () => {
    setContextMenu({});
  };

  return (
    <div onMouseDown={eventHandler}> // attach listener within component
    // ...
```

**Reply by ChristianMurphy** [reply]
Taking an even further step back.
Probably don't add an event listener at all. 🙂 
You are managing state through your router.
Have the router update the content menu.

**Reply by hogoww** [reply]
> > Still, in the link, the anchor does not receive any event, so you cannot click on it.
> 
> There's a lot going wrong here, and there is no framework bug, but there is a good learning opportunity.
> [...]

Oh, so this is expected behavior !
Just a very convoluted interaction.
Cool, I learned a lot here, thank you !

> In addition you have several other bugs, including not cleaning up your event listeners on re-render and not having the function inside the effect. If you choose to keep a document wide listener it should look like:

To be honest, I was just exhausted by the debug session, and just wanted to share my understanding there.
Theses are only in the example, and not in my code base :)

Thanks a bunch for everything @ChristianMurphy that was super interesting !
